packages/api-server/src/methods/normalizer.ts

```diff
@@ -30,10 +30,18 @@ export class EthNormalizer {
     // otherwise read operation might fail the balance check.
     const gasPrice = txCallObj.gasPrice || "0x0";
 
-    // TODO: set default gas limit to min(maxBlockGas, userBalanceAvailableGas)
+    // set default gas limit to min(maxBlockGas, userBalanceAvailableGas)
+    // TODO: use real blockAvailableGas to replace POLY_MAX_BLOCK_GAS_LIMIT
     const maxBlockGasLimit =
       "0x" + BigInt(POLY_MAX_BLOCK_GAS_LIMIT).toString(16);
-    const gas = txCallObj.gas || maxBlockGasLimit;
+    const defaultGasLimit =
+      +gasPrice === 0
+        ? maxBlockGasLimit
+        : min(
+            maxBlockGasLimit,
+            await getMaxGasByBalance(this.rpc, fromAddress, gasPrice)
+          );
+    const gas = txCallObj.gas || defaultGasLimit;
```

Comment:
```suggestion
    let gas = txCallObj.gas;
    if (gas == null) {
        gas = +gasPrice === 0
            ? maxBlockGasLimit
            : min(
                maxBlockGasLimit,
                await getMaxGasByBalance(this.rpc, fromAddress, gasPrice)
              );
    }
```

It is unnecessary to make a `getMaxGasByBalance` call if  `txCallObj.gas` is not `null`